eureka/S3_data_reduction/background.py

```diff
@@ -351,20 +360,174 @@ def fitbg2(dataim, meta, mask, bgmask, deg=1, threshold=5, isrotate=False, isplo
     return bg, bgmask#, mask #,variance
 
 
-def fitbg3(data, isplots=0):
+
+def bkg_sub(img, mask, sigma=5, bkg_estimator='median',
+            box=(10,2), filter_size=(1,1)):
+    """
+    Completes a step for fitting a 2D background
+    model.
+    
+    Parameters
+    ----------
+    img : np.ndarray
+       Single exposure frame.
+    mask : np.ndarray
+       Mask to remove the orders.
+    sigma : float, optional
+       Sigma to remove above. Default is 5.
+    bkg_estimator : str, optional
+       Which type of 2D background model to use.
+       Default is `median`.
+    box : tuple, optional
+       Box size by which to smooth over. Default
+       is (10,2) --> prioritizes smoothing by
+       column.
+    filter_size : tuple, optional
+       The window size of the 2D filter to apply to the
+       low-resolution background map. Default is (1,1).
+
+    Returns
+    -------
+    background : np.ndarray
+       The modeled background image.
+    """
+    sigma_clip = SigmaClip(sigma=sigma)
+    
+    if bkg_estimator.lower()=='mmmbackground':
+        bkg = MMMBackground()
+    elif bkg_estimator.lower()=='median':
+        bkg = MedianBackground()
+        
+    b = Background2D(img, box,
+                     filter_size=filter_size,
+                     bkg_estimator=bkg,                                                                                                                                               
+                     sigma_clip=sigma_clip, fill_value=0.0,
+                     mask=mask)
+    return b.background
+
+
+def skewed_gaussian(x, eta=0, omega=1, alpha=0,scale=1):
+    """ 
+    A skewed gaussian model.
+    """
+    t = alpha * (x - eta) / omega
+    Psi = 0.5 * (1 + erf(t / np.sqrt(2)))
+    psi = 2.0 / (omega * np.sqrt(2 * np.pi)) * np.exp(- (x-eta)**2 / (2.0 * omega**2))
+    return (psi * Psi)*scale
+
+
+def gauss_removal(img, mask, linspace, where='bkg'):
```

Comment:
I was debating whether or not along the time axis would be better! Should've gone with my gut. I wonder if it's worthwhile to split all of these routines into their own `cosmic_ray_removal.py` file. I think there was interest in optimizing CR removal from the Google doc Monday, so it would be cool if people were able to implement their own routines there. What do you think?

Reply:
Yes, 100% agree.
Also, check out clipping.py in the S3 folder.

Reply:
Sorry, I don't see a `clipping.py`? Was it possibly looped into `optspex.py`?

Reply:
See Taylor's current PR

Reply:
Sorry, S4

Reply:
Ooo yeah I see now. I think it makes sense to have the bulk of cosmic ray removal in S3. So maybe we should think about where this step makes the most sense for all instruments, and keep `clipping.py` for nirspec as an additional check for S4

Reply:
If `clipping.py` is generic enough, we could move it to the `lib` directory and use it for multiple stages.  It shouldn't matter if we're looking at the white LC flux or flux along a single pixel in time, the code effectively does the same job.  What do you think?

Reply:
I like this! I think it makes a lot of sense to move it to `lib` since it is pretty generic. I will wait until Taylor's PR is accepted before making any changes, to avoid potential merge issues

Reply:
PR merged!

Reply:
Moved `clipping.py` into `lib` and updated the import statements calling that file! And put my cosmic ray removal function in there